hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java

```diff
@@ -496,7 +537,7 @@ protected boolean shouldCompleteWithoutBackups() {
             // the consequence would be that the backups are never be made and the effects of the map.put() will never be visible,
             // even though the future returned a value;
             // so if we would complete the future, a response is sent even though its changes never made it into the system
-            resetAndReInvoke();
+            context.invocationMonitor.execute(new InvocationRetryTask(true));
```

Comment:
- Yes additional litter is created and it's not absolutely necessary. However I would argue that this is extremely stray path and having this litter cleans up the code. We already create a new object for each retry we do for invocations in `handleRetry()`. I would say the code is as it is only because of historical reasons. But I will change it to the older version if you suggest so. Note that even if I change to the older version, I would still need to create litter (`InvocationRetryTask`) if `invocationCanRetry` is false as I would need to schedule something later on. Conditions must be met in order to reach here:
  * It must be `Invocation`, not `ClientInvocation`.
  * Invocation itself must succeed and backup should timeout.
  * And the member which received the backup operation must left the cluster.

- This is intentional but open to discussion. `async` parameter only effects if the invocation is local. AFAIU, we want to execute operation in current thread to avoid context switching costs. However, `InvocationMonitorThread` should never block IMO. Assume what would happen if a retried invocation is computationally expensive or blocked for some reason. All `InvocationMonitor` would be rendered useless. IMO we should never execute any operations on `InvocationMonitorThread ` as it's dangerous to occupy this thread. Also note that none of this matters because `InvocationMonitorThread` marked `OperationHostileThread`. So even if we run with `async: false`, it will be executed, not run.